File: RizPad/RizPad/TouchKeyMapper.swift
//
//  TouchKeyMapper.swift
//  RizPad
//
//  Slot-based key assignment with cooldown lock:
//  - Each new finger gets the **lowest available slot** (key index).
//  - Moving the finger anywhere on screen does NOT change its key.
//  - Lifting the finger releases the key and frees the slot.
//  - Supports up to 10 simultaneous fingers.
//
//  Cooldown lock:
//  - When a slot is released, it enters a cooldown period (default 500 ms).
//  - During cooldown, the slot is LOCKED and will be skipped when
//    assigning new touches.
//  - After cooldown expires, the slot becomes available again.
//  - ALL recently-released slots can be in cooldown simultaneously.
//
//  This naturally achieves rapid-tap alternation:
//    tap → slot 0 (A), release → slot 0 locked
//    tap again (<500ms) → slot 0 locked, → slot 1 (S), release → slot 1 locked
//    tap again (<500ms) → slot 1 locked, slot 0 unlocked → slot 0 (A)
//    ...
//

import UIKit

/// Manages slot-based touch → key assignment.
final class TouchKeyMapper {

    // MARK: - Configuration

    /// Ordered key list. Index 0 = first slot, index 1 = second slot, etc.
    let keys: [UInt8]

    /// After a slot is released, it stays locked for this duration (seconds).
    /// Set to 0 to disable cooldown.
    var cooldownDuration: TimeInterval = 0.5

    // MARK: - Default layout

    /// Extended key pool — enough slots so that cooldown locks never
    /// run out of available keys even under rapid multi-finger tapping.
    /// Primary 10 keys + 16 overflow keys = 26 total.
    static let defaultKeys: [UInt8] = [
        // Primary (rhythm-game row)
        UInt8(ascii: "a"), UInt8(ascii: "s"), UInt8(ascii: "d"), UInt8(ascii: "f"), UInt8(ascii: "g"),
        UInt8(ascii: "h"), UInt8(ascii: "j"), UInt8(ascii: "k"), UInt8(ascii: "l"), UInt8(ascii: ";"),
        // Overflow (bottom row → top row)
        UInt8(ascii: "z"), UInt8(ascii: "x"), UInt8(ascii: "c"), UInt8(ascii: "v"), UInt8(ascii: "b"),
        UInt8(ascii: "n"), UInt8(ascii: "m"), UInt8(ascii: "q"), UInt8(ascii: "w"), UInt8(ascii: "e"),
        UInt8(ascii: "r"), UInt8(ascii: "t"), UInt8(ascii: "y"), UInt8(ascii: "u"), UInt8(ascii: "i"),
        UInt8(ascii: "o"), UInt8(ascii: "p"),
    ]

    init(keys: [UInt8] = TouchKeyMapper.defaultKeys) {
        self.keys = keys
    }

    // MARK: - Slot tracking

    /// Maps an active UITouch → assigned slot index.
    private(set) var touchSlots: [UITouch: Int] = [:]

    /// Which slot indices are currently occupied by a finger.
    private var occupiedSlots: Set<Int> = []

    /// Timestamp when each slot was last released (for cooldown).
    private var slotReleaseTime: [Int: TimeInterval] = [:]

    // MARK: - Public API

    /// Assign the lowest available (not occupied AND not in cooldown) slot
    /// to a new touch.  Returns the slot index, or `nil` if all slots are
    /// full or locked.
    @discardableResult
    func assignSlot(for touch: UITouch) -> Int? {
        // If already assigned (shouldn't happen), return existing
        if let existing = touchSlots[touch] { return existing }

        let now = ProcessInfo.processInfo.systemUptime

        // Find lowest free slot that is NOT in cooldown
        for i in 0..<keys.count {
            if !occupiedSlots.contains(i) && !isInCooldown(slot: i, now: now) {
                touchSlots[touch] = i
                occupiedSlots.insert(i)
                return i
            }
        }

        // Fallback: if all free slots are in cooldown, pick the one whose
        // cooldown is closest to expiring (least recently released).
        var bestSlot: Int? = nil
        var earliestRelease: TimeInterval = .greatestFiniteMagnitude
        for i in 0..<keys.count {
            if !occupiedSlots.contains(i) {
                let rt = slotReleaseTime[i] ?? 0
                if rt < earliestRelease {
                    earliestRelease = rt
                    bestSlot = i
                }
            }
        }
        if let slot = bestSlot {
            touchSlots[touch] = slot
            occupiedSlots.insert(slot)
            return slot
        }

        return nil // truly all slots occupied
    }

    /// Release the slot for a touch that ended/cancelled.
    func releaseSlot(for touch: UITouch) {
        if let slot = touchSlots.removeValue(forKey: touch) {
            occupiedSlots.remove(slot)
            slotReleaseTime[slot] = ProcessInfo.processInfo.systemUptime
        }
    }

    /// The key (ASCII byte) for a given touch, or `nil`.
    func key(for touch: UITouch) -> UInt8? {
        guard let slot = touchSlots[touch], slot < keys.count else { return nil }
        return keys[slot]
    }

    /// The slot index assigned to a given touch, or `nil`.
    func slot(for touch: UITouch) -> Int? {
        return touchSlots[touch]
    }

    /// Returns the sorted list of ASCII key codes currently pressed.
    func pressedKeys() -> [UInt8] {
        return occupiedSlots.sorted().compactMap { slot in
            slot < keys.count ? keys[slot] : nil
        }
    }

    /// Key label string for a slot index.
    func keyLabel(for slot: Int) -> String {
        guard slot < keys.count else { return "?" }
        return String(UnicodeScalar(keys[slot]))
    }

    /// Primary key label (same as keyLabel now — no alternation).
    func primaryKeyLabel(for slot: Int) -> String {
        return keyLabel(for: slot)
    }

    // MARK: - Private

    /// Whether a slot is currently in its post-release cooldown.
    private func isInCooldown(slot: Int, now: TimeInterval) -> Bool {
        guard let releaseTime = slotReleaseTime[slot] else { return false }
        return (now - releaseTime) < cooldownDuration
    }
}

File: RizPad-Android/app/src/main/java/com/rizpad/touch/TouchKeyMapper.kt
package com.rizpad.touch

import android.os.SystemClock

/**
 * Slot-based key assignment with cooldown lock (Android port of iOS TouchKeyMapper).
 *
 * - Each new finger (pointer ID) gets the **lowest available slot** (key index).
 * - Moving the finger anywhere on screen does NOT change its key.
 * - Lifting the finger releases the key and frees the slot.
 * - Supports up to 10 simultaneous fingers.
 *
 * Cooldown lock:
 * - When a slot is released, it enters a cooldown period (default 500 ms).
 * - During cooldown, the slot is LOCKED and will be skipped.
 * - After cooldown expires, the slot becomes available again.
 */
class TouchKeyMapper(
    val keys: List<Byte> = DEFAULT_KEYS,
    var cooldownDuration: Long = 500L // milliseconds
) {

    companion object {
        val DEFAULT_KEYS: List<Byte> = listOf(
            // Primary (rhythm-game row)
            'a'.code.toByte(), 's'.code.toByte(), 'd'.code.toByte(), 'f'.code.toByte(), 'g'.code.toByte(),
            'h'.code.toByte(), 'j'.code.toByte(), 'k'.code.toByte(), 'l'.code.toByte(), ';'.code.toByte(),
            // Overflow (bottom row → top row)
            'z'.code.toByte(), 'x'.code.toByte(), 'c'.code.toByte(), 'v'.code.toByte(), 'b'.code.toByte(),
            'n'.code.toByte(), 'm'.code.toByte(), 'q'.code.toByte(), 'w'.code.toByte(), 'e'.code.toByte(),
            'r'.code.toByte(), 't'.code.toByte(), 'y'.code.toByte(), 'u'.code.toByte(), 'i'.code.toByte(),
            'o'.code.toByte(), 'p'.code.toByte(),
        )

        const val PRIMARY_SLOT_COUNT = 10
    }

    /** Pointer ID → assigned slot index */
    val pointerSlots: MutableMap<Int, Int> = mutableMapOf()

    private val occupiedSlots: MutableSet<Int> = mutableSetOf()
    private val slotReleaseTime: MutableMap<Int, Long> = mutableMapOf()

    /**
     * Assign the lowest available slot to the given pointer ID.
     * Returns the slot index, or null if all slots are occupied.
     */
    fun assignSlot(pointerId: Int): Int? {
        pointerSlots[pointerId]?.let { return it }
        val now = SystemClock.elapsedRealtime()

        // Find lowest available slot not in cooldown
        for (i in keys.indices) {
            if (i !in occupiedSlots && !isInCooldown(i, now)) {
                pointerSlots[pointerId] = i
                occupiedSlots.add(i)
                return i
            }
        }

        // Fallback: pick slot with closest-to-expiring cooldown
        var bestSlot: Int? = null
        var earliestRelease = Long.MAX_VALUE
        for (i in keys.indices) {
            if (i !in occupiedSlots) {
                val rt = slotReleaseTime[i] ?: 0L
                if (rt < earliestRelease) {
                    earliestRelease = rt
                    bestSlot = i
                }
            }
        }

        if (bestSlot != null) {
            pointerSlots[pointerId] = bestSlot
            occupiedSlots.add(bestSlot)
            return bestSlot
        }

        return null
    }

    /** Release the slot occupied by the given pointer ID. */
    fun releaseSlot(pointerId: Int) {
        val slot = pointerSlots.remove(pointerId) ?: return
        occupiedSlots.remove(slot)
        slotReleaseTime[slot] = SystemClock.elapsedRealtime()
    }

    /** Get the key byte for a given pointer ID. */
    fun key(pointerId: Int): Byte? {
        val slot = pointerSlots[pointerId] ?: return null
        return if (slot < keys.size) keys[slot] else null
    }

    /** Get the slot index for a given pointer ID. */
    fun slot(pointerId: Int): Int? = pointerSlots[pointerId]

    /** Get all currently pressed key bytes, sorted by slot. */
    fun pressedKeys(): List<Byte> {
        return occupiedSlots.sorted().mapNotNull { slot ->
            if (slot < keys.size) keys[slot] else null
        }
    }

    /** Get the display label for a slot. */
    fun keyLabel(slot: Int): String {
        if (slot >= keys.size) return "?"
        return String(charArrayOf(keys[slot].toInt().toChar()))
    }

    /** Get the display label for a primary slot. */
    fun primaryKeyLabel(slot: Int): String = keyLabel(slot)

    private fun isInCooldown(slot: Int, now: Long): Boolean {
        val releaseTime = slotReleaseTime[slot] ?: return false
        return (now - releaseTime) < cooldownDuration
    }
}
